src/BaseServer.ts

```diff
@@ -1,177 +1,70 @@
-        this.identifierName = identifierName;
-        this.defaultQuery = defaultQuery;
-
-        if (data) {
-            this.init(data);
-        }
-    }
-
-    /**
-     * Shortcut for adding several collections if identifierName is always the same
-     */
-    init(data: Record<string, CollectionItem[] | CollectionItem>) {
-        for (const name in data) {
-            const value = data[name];
-            if (Array.isArray(value)) {
-                this.addCollection(
-                    name,
-                    new Collection({
-                        items: value,
-                        identifierName: this.identifierName,
-                        getNewId: this.getNewId,
-                    }),
-                );
-            } else {
-                this.addSingle(name, new Single(value));
-            }
-        }
-    }
-
-    toggleLogging() {
-        this.loggingEnabled = !this.loggingEnabled;
-    }
-
-    /**
-     * @param Function ResourceName => object
-     */
-    setDefaultQuery(query: QueryFunction) {
-        this.defaultQuery = query;
-    }
-
-    setBatchUrl(batchUrl: string) {
-        this.batchUrl = batchUrl;
-    }
-
-    /**
-     * @deprecated use setBatchUrl instead
-     */
-    setBatch(url: string) {
-        console.warn(
-            'Server.setBatch() is deprecated, use Server.setBatchUrl() instead',
-        );
-        this.batchUrl = url;
-    }
-
-    addCollection<T extends CollectionItem = CollectionItem>(
-        name: string,
-        collection: Collection<T>,
-    ) {
-        this.collections[name] = collection;
-        collection.setServer(this);
-        collection.setName(name);
-    }
-
-    getCollection(name: string) {
-        return this.collections[name];
-    }
-
-    getCollectionNames() {
-        return Object.keys(this.collections);
+import {
+    type BaseResponse,
+    type FakeRestContext,
+    AbstractBaseServer,
+} from './AbstractBaseServer.js';
+
+export abstract class BaseServer<
+    RequestType,
+    ResponseType,
+> extends AbstractBaseServer {
+    middlewares: Array<Middleware<RequestType>> = [];
+
+    extractContext(
+        request: RequestType,
+    ): Promise<
+        Pick<FakeRestContext, 'url' | 'method' | 'params' | 'requestJson'>
+    > {
+        throw new Error('Not implemented');
     }
 
-    addSingle<T extends CollectionItem = CollectionItem>(
-        name: string,
-        single: Single<T>,
-    ) {
-        this.singles[name] = single;
-        single.setServer(this);
-        single.setName(name);
+    respond(
+        response: BaseResponse | null,
+        request: RequestType,
+        context: FakeRestContext,
+    ): Promise<ResponseType> {
+        throw new Error('Not implemented');
     }
 
-    getSingle(name: string) {
-        return this.singles[name];
-    }
+    async handle(request: RequestType): Promise<ResponseType | undefined> {
+        const context = this.getContext(await this.extractContext(request));
```

Comment:
Something is wrong with the naming of these methods. You use the name "context" twice to designate two different things. 

I suggest something like:

```jsx
const context = this.getContext(await this.getNormalizedRequest(request));
```